File: platform/Apple/Sources/SweetLineCoreInternal/api/TextAnalyzer.swift
import Foundation
import SweetLineBridge

public final class TextAnalyzer {
    fileprivate let nativeHandle: sl_analyzer_handle_t

    init(nativeHandle: sl_analyzer_handle_t) {
        self.nativeHandle = nativeHandle
    }

    public func analyze(_ text: String) throws -> DocumentHighlight {
        try NativeInterop.withCString(text) { cText in
            let result = sl_text_analyze(nativeHandle, cText)
            defer {
                if let result {
                    sl_free_buffer(result)
                }
            }
            return try HighlightResultDecoder.decodeDocumentHighlight(from: result)
        }
    }

    public func analyzeLine(_ text: String, info: TextLineInfo) throws -> LineAnalyzeResult {
        let data = try NativeInterop.withCString(text) { cText in
            try NativeInterop.withInt32Array([info.line, info.startState, info.startCharOffset]) { lineInfo in
                let result = sl_text_analyze_line(nativeHandle, cText, lineInfo)
                return try NativeInterop.copyInt32Prefix(from: result, initialCount: 5, totalCount: NativeInterop.countLineAnalyzeInts)
            }
        }
        return try HighlightResultDecoder.decodeLineAnalyzeResult(from: data)
    }

    public func analyzeIndentGuides(_ text: String) throws -> IndentGuideResult {
        try NativeInterop.withCString(text) { cText in
            let result = sl_text_analyze_indent_guides(nativeHandle, cText)
            defer {
                if let result {
                    sl_free_buffer(result)
                }
            }
            return try IndentGuideDecoder.decode(from: result)
        }
    }
}

File: platform/Java22/sweetline/src/main/java/com/qiplat/sweetline/TextAnalyzer.java
package com.qiplat.sweetline;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;

import static java.lang.foreign.ValueLayout.JAVA_INT;

/**
 * Plain text highlight analyzer, no incremental update support.
 * Suitable for full text analysis or single-line analysis scenarios.
 * <p>
 * Implements {@link AutoCloseable} for deterministic native resource release.
 */
public class TextAnalyzer implements AutoCloseable {

    private final MemorySegment handle;
    private boolean closed = false;

    TextAnalyzer(MemorySegment handle) {
        this.handle = handle;
    }

    /**
     * Analyze a text and return the highlight result for the entire text.
     *
     * @param text Full text content
     * @return Highlight result
     */
    public DocumentHighlight analyzeText(String text) {
        ensureOpen();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment textSeg = arena.allocateFrom(text);
            MemorySegment resultPtr = (MemorySegment) SweetLineNative.sl_text_analyze.invoke(handle, textSeg);
            if (resultPtr.equals(MemorySegment.NULL)) {
                return new DocumentHighlight();
            }
            try {
                return BufferParser.readDocumentHighlight(resultPtr);
            } finally {
                SweetLineNative.sl_free_buffer.invoke(resultPtr);
            }
        } catch (Throwable e) {
            throw new RuntimeException("Failed to analyze text", e);
        }
    }

    /**
     * Analyze a single line of text.
     *
     * @param text Single line text content
     * @param info Metadata for the current line
     * @return Single line analysis result
     */
    public LineAnalyzeResult analyzeLine(String text, TextLineInfo info) {
        ensureOpen();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment textSeg = arena.allocateFrom(text);
            // Pack TextLineInfo into int32_t[3]
            MemorySegment lineInfoSeg = arena.allocate(JAVA_INT, 3);
            lineInfoSeg.setAtIndex(JAVA_INT, 0, info.line());
            lineInfoSeg.setAtIndex(JAVA_INT, 1, info.startState());
            lineInfoSeg.setAtIndex(JAVA_INT, 2, info.startCharOffset());

            MemorySegment resultPtr = (MemorySegment) SweetLineNative.sl_text_analyze_line
                    .invoke(handle, textSeg, lineInfoSeg);
            if (resultPtr.equals(MemorySegment.NULL)) {
                return new LineAnalyzeResult(new LineHighlight(), 0, 0);
            }
            try {
                return BufferParser.readLineAnalyzeResult(resultPtr, info.line());
            } finally {
                SweetLineNative.sl_free_buffer.invoke(resultPtr);
            }
        } catch (Throwable e) {
            throw new RuntimeException("Failed to analyze line", e);
        }
    }

    /**
     * Perform indent guide analysis on a text.
     *
     * @param text Full text content
     * @return Indent guide analysis result
     */
    public IndentGuideResult analyzeIndentGuides(String text) {
        ensureOpen();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment textSeg = arena.allocateFrom(text);
            MemorySegment resultPtr = (MemorySegment) SweetLineNative.sl_text_analyze_indent_guides
                    .invoke(handle, textSeg);
            if (resultPtr.equals(MemorySegment.NULL)) {
                return new IndentGuideResult();
            }
            try {
                return BufferParser.readIndentGuideResult(resultPtr);
            } finally {
                SweetLineNative.sl_free_buffer.invoke(resultPtr);
            }
        } catch (Throwable e) {
            throw new RuntimeException("Failed to analyze indent guides", e);
        }
    }

    @Override
    public void close() {
        if (!closed) {
            closed = true;
            // TextAnalyzer handle is managed by the engine internally;
            // the C API does not expose a separate free function for text analyzer.
            // The handle is freed when the engine is destroyed.
        }
    }

    private void ensureOpen() {
        if (closed) {
            throw new IllegalStateException("TextAnalyzer is already closed");
        }
    }
}
